File: app/src/main/java/ru/mammoth70/wherearetheynow/LocationActivity.java
package ru.mammoth70.wherearetheynow;

import android.content.Context;
import android.content.Intent;
import android.widget.Button;
import android.widget.PopupMenu;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import java.util.Objects;

public abstract class LocationActivity extends AppCompatActivity {
    // Абстрактный класс для создания Activity вывода геолокации, переданной через intent.

    protected Intent intent;
    protected PointRecord startRecord;
    protected TextView tvName;
    protected TextView tvDateTime;
    protected Button menuButton;

    protected void createFrameTitle(Context context) {
        // Метод вызывается при создании Activity.
        // Не должен переопределяться, но должен вызываться из onCreate после вызова setContentView.
        intent = getIntent();
        tvName = findViewById(R.id.tvName);
        tvDateTime = findViewById(R.id.tvDateTime);
        startRecord = new PointRecord(
                intent.getStringExtra(Util.INTENT_EXTRA_SMS_FROM),
                intent.getDoubleExtra(Util.INTENT_EXTRA_LATITUDE,0),
                intent.getDoubleExtra(Util.INTENT_EXTRA_LONGITUDE,0),
                intent.getStringExtra(Util.INTENT_EXTRA_TIME));
        tvName.setText(Util.phone2name.get(startRecord.phone));
        tvDateTime.setText(MapUtil.timePassed(startRecord.datetime,context));

        // Настроить вызов меню со списком контактов.
        Util.menuPhones.clear();
        try  (DBhelper dbHelper = new DBhelper(context)) {
            dbHelper.getMenuUsers();
        }
        menuButton = findViewById(R.id.btnMenuUsers);
        menuButton.setOnClickListener(view -> {
            PopupMenu popupMenu = new PopupMenu(context, view);
            for (String key : Util.menuPhones) {
                if ((Util.phone2record.containsKey(key)) && (Util.phone2id.containsKey(key))) {
                    Integer id = Util.phone2id.get(key);
                    if (id != null) {
                        popupMenu.getMenu().add(0, id, 0, Util.phone2name.get(key));
                    }
                }
            }
            popupMenu.setOnMenuItemClickListener(item -> {
                reloadMapFromId(context, item.getItemId());
                return true;
            });
            popupMenu.show(); // Отобразить меню
        });
    }

    protected void reloadMapFromId(Context context, int id) {
        // Метод выводит координаты тектом, меняя заголовок карты.
        // Вызывается из меню со списком контактов.
        // После выполнения, вызывает метод reloadMapFromPoint.
        // Может быть переопределён.
        String phone = Util.id2phone.get(id);
        tvName.setText(Util.phone2name.get(phone));
        tvDateTime.setText(MapUtil.timePassed(
                Objects.requireNonNull(Util.phone2record.get(phone)).datetime,context));
        reloadMapFromPoint(context, Objects.requireNonNull(Util.phone2record.get(phone)));
    }

    protected abstract void reloadMapFromPoint(Context context, PointRecord rec);
    // Абстрактный класс, должен быть переопределён.
    // Вызывается из reloadMapFromPoint, а также из OnCreate.
    // Метод перестраивает карту по передаваемой записи PoinRecord.
}

File: app/src/main/java/ru/mammoth70/wherearetheynow/LocationActivity.kt
package ru.mammoth70.wherearetheynow

import android.content.Context
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.PopupMenu
import com.google.android.material.appbar.MaterialToolbar

abstract class LocationActivity : AppActivity() {
    // Абстрактный класс для создания Activity вывода геолокации, переданной через intent.


    protected val startRecord: PointRecord by lazy { PointRecord(
        intent.getStringExtra(INTENT_EXTRA_SMS_FROM)!!,
        intent.getDoubleExtra(INTENT_EXTRA_LATITUDE, 0.0),
        intent.getDoubleExtra(INTENT_EXTRA_LONGITUDE, 0.0),
        intent.getStringExtra(INTENT_EXTRA_TIME)!!) }

    override val topAppBar: MaterialToolbar by lazy { findViewById(R.id.topAppBarMap) }


    override fun onCreate(savedInstanceState: Bundle?) {
        // Функция вызывается при создании Activity.
        // Может быть переопределена, но обычно не переопределяется.

        super.onCreate(savedInstanceState)

        createFrameTitle(this)
        initMap(this)
        reloadMapFromPoint(this, startRecord)
    }


    protected fun createFrameTitle(context: Context) {
        // Функция вызывается из onCreate после вызова setContentView.
        // Не может переопределяться.
        // Функция выводит в заголовок карты имя контакта и время получения координат из записи startRecord,
        // а также создаёт меню со списком контактов в панели заголовка.

        topAppBar.setTitle(DataRepository.getUser(startRecord.phone)?.name)
        topAppBar.setSubtitle(timePassed(startRecord.dateTime, context))
        topAppBar.setNavigationOnClickListener {
            finish()
        }

        topAppBar.setOnMenuItemClickListener { menuItem ->
            when (menuItem.itemId) {
                R.id.itemUsers -> {
                    val anchorView = findViewById<View>(R.id.itemUsers)

                    // Если вдруг не нашли (бывает в редких случаях), используем сам тулбар
                    val popupMenu = PopupMenu(context, anchorView ?: topAppBar)
                    popupMenu.gravity = Gravity.END

                    val menuMapping = mutableMapOf<Int, Long>() // Карта соответствия id и пункта меню.

                    DataRepository.menuPhones.forEachIndexed { index, phone ->
                        DataRepository.getUser(phone)?.let { user ->
                            menuMapping[index] = user.id
                            popupMenu.menu.add(0, index, 0, user.name)
                        }
                    }

                    popupMenu.setOnDismissListener {
                        menuMapping.clear()
                    }

                    popupMenu.setOnMenuItemClickListener { item ->
                        val userId = menuMapping[item.itemId]
                        userId?.let { reloadMapFromId(context, userId) }
                        true
                    }

                    popupMenu.show()
                    true
                }

                else -> false
            }
        }
    }


    protected open fun reloadMapFromId(context: Context, id: Long) {
        // Функция вызывается из меню со списком контактов.
        // Может быть переопределена, но обычно не переопределяется.
        // Функция выводит в заголовок карты имя контакта и время получения координат.
        // После выполнения, вызывает функцию reloadMapFromPoint.

        val user = DataRepository.getUser(id) ?: return
        val record = user.lastRecord ?: return
        topAppBar.setTitle(user.name)
        topAppBar.setSubtitle(timePassed(record.dateTime, context))
        reloadMapFromPoint(context, record)
    }


    protected abstract fun initMap(context: Context)
    // Абстрактная функция, должна быть переопределена.
    // Вызывается из onCreate после createFrameTitle и перед reloadMapFromPoint.
    // Функция делает начальную настройку карты.


    protected abstract fun reloadMapFromPoint(context: Context, rec: PointRecord)
    // Абстрактная функция, должна быть переопределена.
    // Вызывается из onCreate после initMap, вызывается также из reloadMapFromId.
    // Функция перестраивает карту по передаваемой записи PointRecord.

}